drivers/led_strip/ws2812b_sw.c

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <led_strip.h>
+
+#include <string.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_LED_STRIP_LEVEL
+#include <logging/sys_log.h>
+
+#include <zephyr.h>
+#include <board.h>
+#include <gpio.h>
+#include <device.h>
+#include <clock_control.h>
+
+#define BLOCKING ((void *)1)
+
+static int send_buf(u8_t *buf, size_t len)
+{
+	/* Address of OUTSET. OUTCLR is OUTSET + 4 */
+	volatile u32_t *base = (u32_t *)(NRF_GPIO_BASE + 0x508);
+	u32_t pin = BIT(CONFIG_WS2812B_SW_GPIO_PIN);
+	struct device *clock;
+	unsigned int key;
+	/* Initilization of i is strictly not needed, but it avoids an
+	 * uninitialized warning with the inline assembly.
+	 */
+	u32_t i = 0;
+
+	clock = device_get_binding(CONFIG_CLOCK_CONTROL_NRF5_M16SRC_DRV_NAME);
+	if (!clock) {
+		SYS_LOG_ERR("Unable to get HF clock");
+		return -EIO;
+	}
+
+	/* The inline assembly further below is designed to work only with
+	 * the 16 MHz clock enabled.
+	 */
+	clock_control_on(clock, BLOCKING);
+	key = irq_lock();
+
+	while (len--) {
+		u32_t b = *buf++;
+
+		/* Generate signal out of the bits, MSB. 1-bit should be
+		 * roughly 0.85us high, 0.4us low, whereas a 0-bit should be
+		 * roughly 0.4us high, 0.85us low.
+		 */
+		__asm volatile ("movs %[i], #8\n" /* i = 8 */
+				".start_bit:\n"
+
+				/* OUTSET = BIT(LED_PIN) */
+				"strb %[p], [%[r], #0]\n"
+
+				/* if (b & 0x80) goto .long */
+				"tst %[b], %[m]\n"
+				"bne .long\n"
+
+				/* 0-bit */
+				"nop\nnop\n"
+				/* OUTCLR = BIT(LED_PIN) */
+				"strb %[p], [%[r], #4]\n"
+				"nop\nnop\nnop\n"
+				"b .next_bit\n"
+
+				/* 1-bit */
+				".long:\n"
+				"nop\nnop\nnop\nnop\nnop\nnop\nnop\n"
+				/* OUTCLR = BIT(LED_PIN) */
+				"strb %[p], [%[r], #4]\n"
+
+				".next_bit:\n"
+				/* b <<= 1 */
+				"lsl %[b], #1\n"
+				/* i-- */
+				"sub %[i], #1\n"
+				/* if (i > 0) goto .start_bit */
+				"bne .start_bit\n"
+				:
+				[i] "+r" (i)
+				:
+				[b] "l" (b),
+				[m] "l" (0x80),
+				[r] "l" (base),
+				[p] "r" (pin)
+				:);
+	}
+
+	irq_unlock(key);
+	clock_control_off(clock, NULL);
+
+	return 0;
+}
+
+static int ws2812b_sw_update_rgb(struct device *dev, struct led_rgb *pixels,
+				 size_t num_pixels)
+{
+	size_t i;
+
+	/* Convert from RGB to GRB format */
+	for (i = 0; i < num_pixels; i++) {
+		u8_t r = pixels[i].r;
+
+		pixels[i].r = pixels[i].g;
+		pixels[i].g = r;
+	}
+
+	return send_buf((u8_t *)pixels, num_pixels * sizeof(*pixels));
```

Comment:
This cast doesn't work if CONFIG_LED_STRIP_RGB_SCRATCH=y (which could conceivably happen if you want to use an APA102 on the same board). Just to be safe, can you be a bit more defensive?

Reply:
I've added `depends on !CONFIG_LED_STRIP_RGB_SCRATCH` to Kconfig to protect against this happening.

This actually brings me to another efficiency related issue with the current LED API: it'd be nice if we could find some solution which would let the application pass const (ROM) buffers to the API, and have those drivers which don't have the scratch u8_t be able to pass the buffer as-is to the hardware. This wouldn't just be about avoiding translations in the driver, but allowing e.g. bigger animations stored in ROM when there's not enough space in RAM for them.

Reply:
> I've added depends on !CONFIG_LED_STRIP_RGB_SCRATCH to Kconfig to protect against this happening.

Hrm... why not just get it right for both cases? I don't think it's going to blow any resource budgets to do so. I also note that you're relying on the order of r/g/b in the struct earlier in this routine, which also seems unnecessary.

What's wrong with the approach taken by lpd880x_strip_update_rgb(), which makes no assumptions about struct layout or whether scratch is present?

> This actually brings me to another efficiency related issue with the current LED API: it'd be nice if we could find some solution which would let the application pass const (ROM) buffers to the API, and have those drivers which don't have the scratch u8_t be able to pass the buffer as-is to the hardware. This wouldn't just be about avoiding translations in the driver, but allowing e.g. bigger animations stored in ROM when there's not enough space in RAM for them.

I agree this would be very nice, and that the current API is RAM-hungry in this regard. It's possibly tricky to do in general, though. For example, the SPI-based ws2812 driver would require an 8x storage increase to store a "raw" array of SPI data, since it uses a one byte SPI frame for every bit of pixel data.

Do you have any ideas for a general mechanism? Otherwise, perhaps an ioctl()-style interface where individual drivers can specify chip-specific extensions could perhaps serve.